sound/soc/sof/sof-priv.h

```diff
@@ -157,6 +157,14 @@ struct sof_firmware {
 	u32 payload_offset;
 };
 
+enum sof_dai_feature {
+	SOF_DAI_FEATURE_DSP_ONLY, /* access from DSP only */
+	SOF_DAI_FEATURE_DSPLESS,  /* access from host only */
+	SOF_DAI_FEATURE_CHAIN_DMA,/* access from DSP with CHAIN_DMA mode */
```

Comment:
I would re-word the comments and perhaps the SOF_DAI_FEATURE_DSP_ONLY is redundant?
I mean, one can ask if  a DAI can be used in DSPLESS mode and the answer is yes (it can be used in DSPless mode or with DSP mode) or no (it can only be used via DSP).
Similarly, the CHAIN_DMA question can be answered by: yes, ChainDMA can be used on the DAI or it cannot be used?

What would a HDA DAI report if you pass SOF_DAI_FEATURE_DSP_ONLY? It can run in DSPless mode, so it cannot say yes, but what no means? Can it support DSP mode?

Reply:
maybe reword as DSP_ACCESS, HOST_ACCESS, DSP_ACCESS_CHAIN_DMA
The features are not exclusive or required, it's only the hardware restrictions that come into play to prevent one feature from being used on a specific platform

Reply:
@plbossart, I would expect all DAIs to be usable with via DSP, I find defining it redundant, I also understand why you want to have DSP_ACCESS in the enum - to be able to directly use it as an array index.
On the other hand these definitions beg to be a mask?
What about:
```
enum sof_dai_feature {
	SOF_DAI_DSP_ACCESS,,
	SOF_DAI_DSPLESS_ACCESS,
	SOF_DAI_CHAIN_DMA_SUPPORT,
	SOF_DAI_FEATURE_COUNT,
};

BIT(SOF_DAI_DSP_ACCESS)
BIT(SOF_DAI_DSPLESS_ACCESS) / BIT(SOF_DAI_HOST_ACCESS)
BIT(SOF_DAI_CHAIN_DMA_SUPPORT)
```
I would stick with the terms we are using for DSPLess and ChainDMA here.
The callback then can be 'check_dai_feature(sdev, dai, enum sof_dai_feature)`?

But does it make any sense to all of this? We have only one feature that we need the test for: ChainDMA
I think a single `dai_chain_dma_supported(sdev, dai)` or just a single feature definition to test for would be better and if there will be a need it an be extended, but I don't see that.

Reply:
well we have the existing interface_mask with the magic index 0 and 1, it'd be odd to have two ways of checking if an interface is supported in a specific more on a specific platform....

Reply:
@plbossart, what about dropping the DAI type check (in `sof_ipc4_widget_setup_comp_dai()`) for ChainDMA altogether and leave it to the topology writer to only use it on interfaces where it is allowed?
We fail that step currently if the DAI cannot be used with ChainDMA, if we drop that sanity check then we will likely fail a bit later in firmware I guess. Likely the error is going to be different, but if we have proper prints we will know the reason.

Reply:
I don;t think it's a wise idea to trust the topology blindly. I'll be the first one to screw up...

Reply:
That could be but the first time you try that tplg you will figure it out that it is not right and the final topology would be correct?
If we have the check for the DAI it will also fail, so it is fixed before the topology is ready for review.

I'm not sure if we do that or not, but we should print (dev_dbg) that a ChainDMA is used - but that is obvious form the messages we send, so likely no need to do that.

I just fear that we are trying to over engineer this for no benefits and the naming and definition of the check is not that clear. Especially the naming.

Reply:
obsolete, moved to a non-ambiguous 'is_chain_dma_supported' callback